Add faceted type-ahead search capability to CSET web interface

jfrost-mo opened the pull request:
CSET output web interface improvements developed during my MSc project.

GitHub Copilot was used for review and fix suggestions for the prototype python parser, which was then hand ported to JavaScript. That code has been rebased away.

Thank you to @Pierre-siddall for reviewing the JavaScript code.

### Contribution checklist

Aim to have all relevant checks ticked off before merging. See the [developer's guide](https://metoffice.github.io/CSET/contributing/) for more detail.

- [x] Documentation has been updated to reflect change.
- [x] New code has tests, and affected old tests have been updated.
- [x] All tests and CI checks pass.
- [x] Ensured the pull request title is descriptive.
- [x] Conda lock files have been updated if dependencies have changed.
- [x] Attributed any Generative AI, such as GitHub Copilot, used in this PR.
- [x] Marked the PR as ready to review.

jfrost-mo commented:
# TODO for next year:

1. Fix bugs in parser/web interface.
2. Introduce a three column layout, as it is commonly requested.
3. Tweak the CSS to make it look pretty.
4. Rebase the branch into a sensible set of changes. It may want splitting up into a couple pull requests.

jfrost-mo committed:
Add search capability to CSET UI

Additionally included is a conversion of the index to JSON Lines (In
theory it should be more efficient, but in practice we still load the
whole string first. At least it looks nicer in the index file.)

Non-useful entries are removed from the index before it is written, so
it only contains information the client will need.

jfrost-mo committed:
Simplify filter parser language

Removes redundant negative operators and allows facets to be any
literal. Comparisons are now case insensitive.

jfrost-mo committed:
Add initial filtering help

jfrost-mo commented:
<img width="1920" height="885" alt="cset-web-ui-search" src="https://github.com/user-attachments/assets/bd316e67-342d-4d19-a7c3-8a5ad4fdff17" />
Current web UI. Will need some CSS prettification before merging.

jfrost-mo committed:
Filter plot_resolution

jfrost-mo committed:
Add recipe variables as metadata for filtering

jfrost-mo committed:
Remove more undesired facets from the plot index

jfrost-mo committed:
Improve wording of a couple filter examples

jfrost-mo committed:
Improve styling of facet dropdowns and sidebar in web UI

Pierre-siddall commented on src/CSET/cset_workflow/app/finish_website/file/html/script.js:
```diff
@@ -66,87 +443,138 @@ function ensure_dual_frame() {
+// Update query based on facet dropdown value.
+function updateFacetQuery(e) {
+  const facet = e.target.name;
+  const value = e.target.value;
```
Similarly to the previous comment I reckon that the `e` variable could have a slightly more informative name to make it crystal clear what is being operated on.

Pierre-siddall reviewed (commented):
This looks really good James, you've definitely had a good understanding of the lexer and have implemented clever regex to match the different range of search queries a user may input. Equally, you clearly have demonstrated that you know how the DOM functions by construing HTML elements being added to or removed from the DOM via javascript. I only have a couple of tiny improvements here regarding informative variable names in order to help future developers understand the control flow of your javascript by knowing what is being operated on. However, overall I think this is a well justified piece of work, great job :) .

jfrost-mo committed:
Avoid single character variable names in JavaScript

Using more descriptive names clarifies what data is being used.

jfrost-mo committed:
Update expected results for parbaking

We now add the templated variables into the written out recipe.

daflack reviewed (commented):
I've been trying out the new functionality for a couple of hours to really stress test it. Overall, it's a great change and I really like it. The drop down menus make it really easy to use, and the pop-outs are great for organising plots how I want to see them. 

I have a couple of very minor niggles:
 1. sometimes the filtering functionality doesn't work - see as an example. I can't find a consistent way of breaking it the same way though, or at least haven't noticed any patterns.  After a bit more exploration it seems to be linked to whether date or plot category is selected - you can't appear to select both using the dropdowns.
 2. when the drop down help, and drop down menus it does create a very long menu bar so it is then hard to find the search bar so is there a possibility of keeping the search bar in a fixed location/putting a scroll in for different sections?

<img width="547" height="812" alt="Image" src="https://github.com/user-attachments/assets/c027dd0d-0cd3-4ab5-ba1d-bef7bb61f7fa" />

Otherwise it looks and works great, and I know through regular use I'll get use to it really quickly.

jfrost-mo committed:
Add playwright based web interface test

Runs a localhost webserver in a separate thread for serving the content
under test to make it self contained. Playwright is therefore added as a
testing dependency.

jfrost-mo committed:
Optimise performance of web UI for large runs

The primary optimisation here is to batch the addition of elements to
the DOM. This reduces the number of rerenders/layouts triggered and
minimises the number of events fired. This is especially useful as
uBlock Origin watches DOM mutations, which slows things down a lot.

jfrost-mo committed:
Use an exact match when selecting a facet via the dropdown

jfrost-mo committed:
Add sticky search header

Fraetor committed:
Add facet dropdown browser test

Fraetor committed:
Integrate clear query button into search box

Fraetor committed:
Filter out additional facets from index

These facets can only take a single value, are hard to interpret, and
don't allow any selection that can't be done with category instead.

Fraetor committed:
Store diagnostic records in array rather than DOM

This is far more efficient, leading to lower memory usage, and faster
performance. Rendering was still expensive, so to keep it performant the
number of rendered diagnostics is limited to 500.

jfrost-mo committed:
Gitignore web interface testing symlink

I keep accidentally git adding it. This symlink allows testing the web
interface with real data during development.

jfrost-mo committed:
Move too many diagnostics warning to top of list

This makes it more visible while the user is still near the filter
controls.

jfrost-mo commented:
The search still has issues when three filters are selected at once. Is it related to the ANDs not sharing precedence properly, like with the following bracket pairs?

```
(A [B) C]
```

jfrost-mo committed:
Correct stride causing filter Conditions to be skipped

When more than two Conditions were specified separated by implicit ANDs,
Conditions would sometimes not apply. For example the query "a b c"
would check for a and b, but not c. This was due to an incorrect array
index increment inside the and parsing logic, which caused the next
Condition after a pair of implicitly ANDed Conditions to be skipped.

github-actions[bot] committed:
[CI] Update conda lock files

jfrost-mo committed:
Install playwright browser when calling make test-full

This means the browser shall be installed in CI. Playwright won't
redownload if already downloaded, but will update if needed.

jfrost-mo committed:
Update locator for new clear query button

jfrost-mo committed:
Make diagnostic cutoff warning sticky

It now remains visible even when scrolled down, so you know that you
don't have the full list of results. We also use it to display a message
while the diagnostics are loading.

jfrost-mo committed:
Use CSS variables for all colours

This will potentially allow for theming in future, and lets us see the
range of colours we are using to ensure consistency.

jfrost-mo committed:
Use system-ui default font for sidebar

As it is a UI element we should use the UI font. The placeholder/plot
pages have been felt alone, as they contain more prose for which
san-serif is better.

daflack reviewed (approved):
Having had a brief look at the example website provided, and tested the capability that was breaking and not as clear I find the changes really helpful and beneficial. Website looks really clear and easy to navigate with useful factors considered. Great job! (Probably want multiple approving reviews rather than just one for this PR).
